arangod/Graph/PathEnumerator.cpp

```diff
@@ -51,6 +51,29 @@ void PathEnumerator::setStartVertex(arangodb::velocypack::StringRef startVertex)
   TRI_ASSERT(_enumeratedPath.vertices.size() == 1);
 }
 
+bool PathEnumerator::keepEdge(arangodb::graph::EdgeDocumentToken& eid,
+                              arangodb::velocypack::slice edge,
+                              velocypack::StringRef sourceVertex, size_t depth,
+                              size_t cursorId) {
+  VPackSlice e = edge;
+  if (edge.isString()) {
+    e = _opts->cache()->lookupToken(eid);
+  }
+  return keepEdge(e, sourceVertex, depth, cursorId);
+}
+
+bool PathEnumerator::keepEdge(arangodb::velocypack::Slice edge,
+                              arangodb::velocypack::StringRef sourceVertex,
+                              size_t depth, size_t cursorId) {
+  if (_opts->hasEdgeFilter(depth, cursorId) &&
+      !_traverser->edgeMatchesConditions(edge, sourceVertex, depth, cursorId)) {
+    // This edge does not pass the filtering
+    return false;
+  }
+
+  return _opts->destinationCollectionAllowed(edge, sourceVertex);
+}
```

Comment:
There is now a slight pessimization in case there is no edge filter and there are no collection restrictions.

Previously, when there was no edge filter, the following code didn't execute for each edge:
```
            VPackSlice edge = other;
            if (edge.isString()) {
              edge = _opts->cache()->lookupToken(eid);
            }
```
Now this code is executed for every edge, which means it will make things more expensive in case there is no edge filter.
I think we should do the lookupToken call as late as possible again.